distsys/archetypeinterface.go

```diff
@@ -4,85 +4,507 @@ import (
+//			res, err = res.Index(index)
+//			if err != nil {
+//				return
+//			}
+//		}
+//		value, err = res.ReadValue()
+//
+//		return
+//	}
+//}
+
+//func (iface ArchetypeInterface) BranchRead(handle ArchetypeResourceHandle, indices []tla.TLAValue, branchResources BranchResourceMap) (value tla.TLAValue, err error) {
+//	iface.ensureCriticalSectionWith(handle)
+//
+//	res, ok := branchResources[handle]
+//	if ok {
+//		// Handle was in the branch resources
+//		for _, index := range indices {
+//			res, err = res.Index(index)
+//			if err != nil {
+//				return
+//			}
+//		}
+//		value, err = res.ReadValue()
+//		return
+//	} else {
+//		// Handle wasn't in the branch resources
+//		resource := iface.ctx.getResourceByHandle(handle)
+//		res, err = resource.ForkState()
+//
+//		// Put the new forked resource in the branch resources
+//		branchResources[handle] = res
+//
+//		if err != nil {
+//			return
+//		}
+//		for _, index := range indices {
+//			res, err = res.Index(index)
+//			if err != nil {
+//				return
+//			}
+//		}
+//		value, err = res.ReadValue()
+//		return
+//	}
+//}
+
 // NextFairnessCounter returns number [0,ceiling) indicating a non-deterministic branching decision which,
 // given an MPCal critical section being retried indefinitely with no other changes, will try to guarantee that all
 // possible non-deterministic branch choices will be attempted.
-func (iface ArchetypeInterface) NextFairnessCounter(id string, ceiling uint) uint {
+func (iface *ArchetypeInterface) NextFairnessCounter(id string, ceiling uint) uint {
 	return iface.ctx.fairnessCounter.NextFairnessCounter(id, ceiling)
 }
 
+type BranchResourceMap map[ArchetypeResourceHandle]ArchetypeResource
+
+//type branch func(branchResources BranchResourceMap) error
+type branch func(iface *ArchetypeInterface) error
+
+func (iface *ArchetypeInterface) ForkIFace() (*ArchetypeInterface, error) {
+	childContext := &MPCalContext{
+		archetype: iface.ctx.archetype,
+		self:      iface.ctx.self,
+		// Resources need
+		fairnessCounter: iface.ctx.fairnessCounter,
+		jumpTable:       iface.ctx.jumpTable,
+		procTable:       iface.ctx.procTable,
+
+		dirtyResourceHandles: make(map[ArchetypeResourceHandle]bool),
+
+		constantDefns: iface.ctx.constantDefns,
+		allowRun:      true,
+		awaitExit:     make(chan struct{}),
+	}
+
+	childIFace := &ArchetypeInterface{
+		ctx:             childContext,
+		ForkedResources: make(map[ArchetypeResourceHandle]ArchetypeResource),
+		parent:          iface,
+	}
+
+	return childIFace, nil
+}
+
+func (iface *ArchetypeInterface) LinkIFace(childIFace *ArchetypeInterface) error {
+
+	for forkedHandle, forkedResource := range childIFace.ForkedResources {
+		_, ok := iface.ForkedResources[forkedHandle]
+		if !ok {
+			// Parent iface doesn't contain resource so add it in
+			iface.ForkedResources[forkedHandle] = forkedResource
+		} else {
+			// Parent iface had the resource so link its state
+			err := forkedResource.LinkState()
```

Comment:
imo, this method should not exist. I think there's another way, where you just manipulate the maps and nothing else, until commit.